backend/server/importlib/headerSource.py

```diff
@@ -507,6 +507,12 @@ def populate(self, hash):
                     self[i] = self[i].encode("utf-8")
                 except:
                     self[i] = self[i].encode("iso-8859-1")
+            # Workaround for #4614
+            if type(self[i]) == str:
+                try:
+                    self[i].encode("utf-8")
+                except:
+                    self[i] = repr(self[i])
```

Comment:
Can we try to follow the same logic that is described in the comment above this block (utf-8 -> latin-1)? As it is right now, it might be confusing when we look at the comment and the code in 6 months from now.

Since this code is only ran with Python3, "text" and "name" should be `str` and not `bytes`. I'm currently going through the code base fixing these "str type" inconsistencies that we have, if you want I can also take care of this bug fix and include it in my PR (to be created).

Reply:
Let's see who can get to it first... I will try to reproduce the issue again to test.
If I understand you correctly, all I need to do is change the last line to `self[i] = self[i].encode("iso-8859-1")`, correct?
If that is the case, I would just put an `or` in line 505.

Reply:
If I put it to latin, I get `latin-1' codec can't encode character '\udcf8' in position 19: ordinal not in range(256)`. So same story in a different colour.
Will update the structure and comment instead.

Reply:
ok, I managed to find a way:
```
            # Filter out invalid UTF-8 from string.
            if type(self[i]) == str:
                try:
                    self[i].encode("utf-8")
                except:
                    self[i] = repr(self[i]).encode("utf-8")
```
My understanding is that Python 3 strings are always UTF-8. So idea here is to test for a valif UTF-8 string. if not, use `repr`. That was the easiest way I found to avoid errors. End result will still be UETF-8 due to encoding.

Reply:
Strings in Python 3 are UTF-8 in most cases, but don't have to be (https://stackoverflow.com/questions/49991870/python-default-string-encoding). The question here is: where and how is the changelog header read? 

`self[i].encode("utf-8")` is a good way to check if the string is valid UTF-8 as you said, but the second part is not completely correct. `str.encode()` is the translation from `str` to `bytes`. Since we don't want `bytes` for text, the `repr().encode` should just be a `repr`:
```
>>> type(repr(1))
<class 'str'>
>>> type(repr(1).encode())
<class 'bytes'>
```

I've touched some of the surrounding code, I'll link the PR once I have it opened. I wanted to try my changes on Alma, but so far I wasn't able to set it up.

Reply:
@agraul , `repr` was, what I had in my first attempt (unless I misunderstand you).
If you like, I can test those changes.
Note that currently, Uyuni on Alma does not reliably start. There is a lock condition in the Java code on start. It currently appears that you can get past it if you slow down your machine (enable ALL logging) or speed it up to get past it (8CPUs/16GB Ram).

Reply:
I used your https://github.com/uyuni-project/uyuni/issues/4614#issuecomment-1001214116 to test this out myself. For this very case, the line I suggested above does not work.

That got me thinking, `self[i]` is a `str`, so what's the actual cause of the `psycopg2.errors.NotNullViolation` traceback you see without this patch?

Reply:
@agraul, I did some more extensive testing. We seem to be chasing different issues or at least one issue has changed in the meantime (list item 2 below).

First of all, I can confirm that we are only talking about string objects.
L505: `print ( i, " ", type(self[i]))`
returned type `str` for all entries (`i` being `name` and `text` alternatively)


My tests and attempts inadvertently seem to address all the below issues. I guess there is a dependency on the issues.

**Phase: Importing packages to DB:**
1. #4614 message
**`'utf-8' codec can't encode character '\udcf8' in position 19: surrogates not allowed`**
When this message occurs, the package batch content is not imported and the sync completes without any further issues (just the batch not processed). Note that in the issue, this happens in batch 387 and 388.
2. #4614 
**Stack trace with `/usr/lib64/python3.6/site-packages/psycopg2/extras.py`**
I don't get this stack trace anymore. I can't reproduce. there is a chance that this has been modified. Note that in the issue, this happens in batch 400.

**Phase: Linking packages to the channel.**
3. A new error message during the phase: 
```
21:07:04 ======================================
21:07:04 | Channel: centos7-x86_64
21:07:04 ======================================
21:07:04 Sync of channel started.
21:07:07 
21:07:07   Importing comps file a4e2b46586aa556c3b6f814dad5b16db5a669984d66b68e873586cd7c7253301-c7-x86_64-comps.xml.gz.
21:07:07   Renaming non-standard filename a4e2b46586aa556c3b6f814dad5b16db5a669984d66b68e873586cd7c7253301-c7-x86_64-comps.xml.gz to comps.xml.gz.
21:07:07 *** NOTE: Importing comps file for the channel 'centos7-x86_64'. Previous comps will be discarded.
21:07:15 Repo URL: http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=os
21:07:15     Packages in repo:             10072
21:07:15     Packages passed filter rules:  1023
21:07:17     Packages already synced:       1021
21:07:17     Packages to sync:                 2
21:07:17     New packages to download:         2
21:07:17   Downloading packages:
21:07:17 Downloading total 2 files from 1 queues.
21:07:17     1/2 : compat-libcap1-1.10-7.el7.i686.rpm
21:07:17     2/2 : compat-libcap1-1.10-7.el7.x86_64.rpm
21:07:18 
21:07:18   Importing packages to DB:
21:07:18 'utf-8' codec can't encode character '\udcf8' in position 19: surrogates not allowed
21:07:18   Package batch #1 of 1 completed...
21:07:18 
21:07:18   Linking packages to the channel.
21:07:18 Unexpected error: <class 'spacewalk.server.importlib.importLib.InvalidPackageError'>
21:07:18 Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/spacewalk/satellite_tools/reposync.py", line 595, in sync
    ret = self.import_packages(plugin, data['id'], url, is_non_local_repo)
  File "/usr/lib/python3.6/site-packages/spacewalk/satellite_tools/reposync.py", line 1146, in import_packages
    importer.run()
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/importLib.py", line 778, in run
    self.submit()
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/packageImport.py", line 131, in submit
    self.backend.lookupPackages(self.batch, self.checksums)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 834, in lookupPackages
    raise_with_tb(not_found[0], not_found[1])
  File "/usr/lib/python3.6/site-packages/uyuni/common/usix.py", line 77, in raise_with_tb
    raise value
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 826, in lookupPackages
    self.__lookupObjectCollection([package], 'rhnPackage')
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 2506, in __lookupObjectCollection
    raise InvalidPackageError(object, "Could not find object %s in table %s" % (object, tableName))
spacewalk.server.importlib.importLib.InvalidPackageError: Could not find object [<<class 'spacewalk.server.importlib.importLib.IncompletePackage'> instance; attributes={'package_id': None, 'name': 'compat-libcap1', 'epoch': None, 'version': '1.10', 'release': '7.el7', 'arch': 'i686', 'org_id': 1, 'package_size': None, 'last_modified': None, 'md5sum': None, 'channels': {178: 'centos7-x86_64'}, 'checksum_list': None, 'checksum': '7453914fde0bec5a7849548d54f0f922da541493b2e9f2e6a1be19f57956364b', 'checksum_type': 'sha256', 'checksums': {'sha256': '7453914fde0bec5a7849548d54f0f922da541493b2e9f2e6a1be19f57956364b'}, 'name_id': 1746, 'evr_id': 965, 'package_arch_id': 104, 'nevra_id': 2324, 'checksum_id': 141770}] in table rhnPackage
```

I am not sure yet, what to make out of the last bit. Either the package was not imported as assumed in list item 1 or the item name has changed and cant be matched anymore.

So now I am taking your last idea and enhanced it like that. What do you think?
```
            # Filter out invalid UTF-8 from string.
            try:
                self[i].encode("utf-8")
            except:
                self[i] = self[i].encode('utf-8', errors='replace').decode('utf-8')
```

Reply:
Thanks for looking into the problem again and the detailed comment!

1. sounds like a a problem we should address separately, batches not being processed is something that should not be ignored IMO.
2. I don't know why this happens. I'm not an expert on this code, but I can see we have two different parts where the changelog is handled. I think this traceback has to do with an SQL query, which internally tries to convert the `str` to `bytes` and fails. 
3. It sounds plausible that this one is caused by 1.

> So now I am taking your last idea and enhanced it like that (tested working). What do you think?

I like that one! It yields valid UTF-8 with the "expected question mark" if it does not know what else to put there. I would only change the bare `except` to an `except UnicodeEncodeError`, but otherwise I think this is a good solution to handling the encoding.

@meaksh since you've dealt with a number of encoding/decoding bugs in the past, what do you think? Are you aware of other places where we might need an update?

Reply:
Ahhhhhhh, hang on. The way you phrase point 1, it rings a bell: https://github.com/uyuni-project/uyuni/issues/4746

Reply:
> Do we need this try/except block at all? Can't we just always put self[i] = self[i].encode('utf-8', errors='replace').decode('utf-8')?

I think you're right about that.

Reply:
Done and tested.